peps/pep-0771.rst

```diff
@@ -0,0 +1,505 @@
+    ]
+
+Unlike the previous example however, maintainers would not necessarily provide a
+way to do an installation without any extras since it might leave the package in
+an unusable state.
+
+If packages can support e.g. multiple backends at the same time, and some of
+the backends should always be installed, then the dependencies for these must be given
+as required dependencies rather than using the default extras mechanism.
+
+Supporting minimal installations while not always removing default extras
+-------------------------------------------------------------------------
+
+An additional case we consider here is where a package maintainer wants to support
+minimal installations without any extras, but also wants to support having users
+specify additional extras without removing the default one. Essentially, they
+would want:
+
+* ``package[minimal]`` to give an installation without any extras
+* ``package`` to install recommended dependencies (in a ``recommended`` extras)
+* ``package[additional]`` to install both recommended and additional dependencies (in an ``additional`` extras)
+
+This could be achieved with e.g:
+
+.. code-block:: toml
+
+    [project]
+    default-optional-dependencies = [
+        "recommended"
+    ]
+
+    [project.optional-dependencies]
+    minimal = []
+    recommended = [
+        "package1",
+        "package2"
+    ]
+    additional = [
+        "package[recommended]",
+        "package3"
+    ]
+
+The ability for a package to reference itself in the extras is supported by
+existing Python packaging tools.
+
+Teaching package authors
+------------------------
+
+Packages making use of any of the approaches above must ensure that they
+properly document the options available to users in terms of installation.
+
+Reference Implementation
+========================
+
+The following branch contains a modified version of the `flit
+<https://flit.pypa.io/en/stable/>`_ package which recognizes the
+``default-optional-dependencies`` key in ``pyproject.toml`` and adds the
+relevant ``Default-Extra:`` metadata to the build package:
+
+https://github.com/astrofrog/flit/tree/default-extras-pep
+
+In addition, the following branch contains a modified version of `pip
+<https://pip.pypa.io/en/stable/>`_ which implements the correct logic for
+installing extras when default extras are present:
+
+https://github.com/astrofrog/pip/tree/default-extras-pep
+
+Finally, the following repository makes use of the above two branches to
+test out different scenarios and use cases described in this PEP:
+
+https://github.com/astrofrog/pep-771-demo
+
+The implementations above are proof-of-concepts at this time, do not include any
+tests or documentation in the respective packages, and the existing changes have
+not yet been reviewed by the relevant maintainers. Nevertheless, they are
+functional enough to allow for interested maintainers to try these out.
+
+Rejected Ideas
+==============
+
+Syntax for unselecting extras
```

Comment:
Hmm, perhaps "deselect" instead of "unselect"?
```suggestion
Syntax for deselecting extras
```

Deselected is something that was once selected, but is no longer.

Unselected is something that is not selected, and may not have been before.

If so, change the others below as well.